include/seqan3/alignment/pairwise/execution/alignment_executor_two_way.hpp

```diff
@@ -231,112 +241,158 @@ class alignment_executor_two_way
      * \{
      */
 
-    //!\brief Sets the buffer pointer.
-    void setg(buffer_pointer beg, buffer_pointer end)
+    //!\brief Fills pre-assigned buckets (one bucket = results of one alignment) with new alignment results.
+    underflow_status underflow()
     {
-        gptr = beg;
-        egptr = end;
-    }
-
-    //!\brief Refills the buffer with new alignment results.
-    size_t underflow()
-    {
-        if (gptr < egptr)  // Case: buffer not completely consumed
-            return in_avail();
+        if (!is_buffer_empty())  // Not everything consumed yet.
+            return underflow_status::non_empty_buffer;
 
         if (is_eof())  // Case: reached end of resource.
-            return eof;
-
-        // Reset the get pointer.
-        setg(std::ranges::begin(buffer), std::ranges::end(buffer));
+            return underflow_status::end_of_resource;
 
-        using std::get;
+        // Reset the buckets and the bucket iterator
+        reset_buckets();
 
-        // Apply the alignment execution.
-        size_t count = 0;
-        size_t buffer_limit = in_avail();
-
-        for (size_t current_chunk_size = std::ranges::distance(*chunked_resource_it);
-             count < buffer_limit && !is_eof();
-             count += current_chunk_size, gptr += current_chunk_size, ++chunked_resource_it)
+        // Execute the algorithm (possibly asynchronous) and fill the buckets in this pre-assigned order.
+        for (bucket_end = bucket_iterator; bucket_end != bucket_vector.end() && !is_eof(); ++bucket_end, ++resource_it)
```

Comment:
Because I would need to create an additional variable that is only used to set the bucket_end after the iteration. So I can use it directly here and avoid the superfluous copy.